File: vplanapp/src/main/java/xyz/jbapps/vplan/util/VPlanParser.java
package xyz.jbapps.vplan.util;

import android.support.annotation.NonNull;
import android.util.Log;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.List;

import xyz.jbapps.vplan.data.VPlanData;
import xyz.jbapps.vplan.data.VPlanRow;

/**
 * @author Jannik Bamberger (dev.jbamberger@gmail.com)
 */

public final class VPlanParser {

    private static final String TAG = VPlanParser.class.getSimpleName();

    private static final int GRADE_C = 0;
    private static final int HOUR_C = 1;
    private static final int SUBJECT_C = 2;
    private static final int ROOM_C = 3;
    private static final int KIND_C = 4;
    private static final int CONTENT_C = 5;

    @NonNull
    public static VPlanData parse(@NonNull String html, long lastModified) throws ParseException {
        Log.d(TAG, html);

        Document doc = Jsoup.parse(html);

        VPlanData vData = new VPlanData();
        vData.setLastUpdated(lastModified);
        vData.setStatus(readVPlanStatus(html));
        vData.setTitle(readVPlanTitle(doc));
        vData.setMotd(readMotdTable(doc.getElementsByClass("info")));
        vData.setRows(readVPlanTable(doc.getElementsByClass("list").select("tr")));
        return vData;
    }


    @NonNull
    private static String readVPlanStatus(@NonNull String html) throws ParseException {
        try {
            String status = html.split("</head>")[1];
            status = status.split("<p>")[0].replace("\n", "").replace("\r", "");
            return status;
        } catch (ArrayIndexOutOfBoundsException e) {
            throw new ParseException("Could not parse vplan status", e);
        }
    }

    @NonNull
    private static String readVPlanTitle(@NonNull Document vplanDoc) throws ParseException {
        try {
            return vplanDoc.getElementsByClass("mon_title").get(0).getAllElements().get(0).text();
        } catch (Exception e) {
            throw new ParseException("Could not parse vplan title", e);
        }
    }

    @NonNull
    private static String readMotdTable(@NonNull Elements classInfo) throws ParseException {
        try {
            Elements tableRows = classInfo.select("tr");
            StringBuilder motd = new StringBuilder();
            for (Element line : tableRows) {
                Elements cells = line.children().select("td");
                int size = cells.size();
                if (size == 0) {
                    continue;
                }
                boolean highlightRow = size > 1 && cells.first().text().toLowerCase().contains("unterrichtsfrei");
                if (highlightRow) {
                    motd.append("<font color=#FF5252>");
                }
                for (int i = 0; i < size; i++) {
                    Element cell = cells.get(i);
                    motd.append(cell.toString());
                    if (i < size - 2) {
                        motd.append(" | ");
                    }
                }
                if (highlightRow) {
                    motd.append("</font>");
                }
                motd.append("<br />");
            }
            String data = motd.toString();
            Log.d(TAG, data);
            Document dat = Jsoup.parse(data);
            dat.select("html").unwrap();
            dat.select("head").unwrap();
            dat.select("body").unwrap();
            dat.select("td").unwrap();
            data = dat.toString();
            Log.d(TAG, data);
            return data;
        } catch (Exception e) {
            throw new ParseException("Could not parse motd table", e);
        }
    }

    @NonNull
    private static List<VPlanRow> readVPlanTable(@NonNull Elements vPlanTable) throws ParseException {
        try {
            List<VPlanRow> rows = new ArrayList<>();
            for (Element line : vPlanTable) {
                Elements cells = line.children();
                if(cells.select("th").size() == 0) {
                    rows.add(readVPlanTableCells(cells));
                }
            }
            return rows;
        } catch (Exception e) {
            throw new ParseException("Could not parse vplan table", e);
        }
    }

    @NonNull
    private static VPlanRow readVPlanTableCells(Elements cells) throws ParseException {
        try {
            if (cells.size() >= 6) {
                VPlanRow row = new VPlanRow();

                Elements grade = cells.get(GRADE_C).getElementsByTag("span");
                Elements hour = cells.get(HOUR_C).getElementsByTag("span");
                Elements subject = cells.get(SUBJECT_C).getElementsByTag("span");
                Elements room = cells.get(ROOM_C).getElementsByTag("span");
                Elements kind = cells.get(KIND_C).getElementsByTag("span");
                Elements content = cells.get(CONTENT_C).getElementsByTag("span");

                row.setGrade(grade != null && grade.first() != null
                        ? grade.first().html()
                        : cells.get(GRADE_C).text());
                row.setHour(hour != null && hour.first() != null
                        ? hour.first().html()
                        : cells.get(HOUR_C).text());
                row.setContent(content != null && content.first() != null
                        ? content.first().html()
                        : cells.get(CONTENT_C).text());
                row.setSubject(subject != null && subject.first() != null
                        ? subject.first().html()
                        : cells.get(SUBJECT_C).text());
                row.setRoom(room != null && room.first() != null
                        ? room.first().html()
                        : cells.get(ROOM_C).text());
                row.setKind(kind != null && kind.first() != null
                        ? kind.first().html()
                        : cells.get(KIND_C).text());
                row.setOmitted(cells.get(KIND_C).text().toLowerCase().contains("entfall"));
                row.setMarkedNew(cells.get(GRADE_C).attr("style").matches("background-color: #00[Ff][Ff]00"));

                return row;
            }
            throw new ParseException("Could not parse row, invalid format.");
        } catch (Exception e) {
            throw new ParseException("Could not parse vplan row", e);
        }
    }
}

File: fhg-repository/src/main/java/de/jbamberger/fhg/repository/api/VPlanParser.kt
package de.jbamberger.fhg.repository.api

import androidx.annotation.VisibleForTesting
import de.jbamberger.fhg.repository.api.VPlanParser.Columns.*
import de.jbamberger.fhg.repository.data.VPlanDay
import de.jbamberger.fhg.repository.data.VPlanHeader
import de.jbamberger.fhg.repository.data.VPlanRow
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.ResponseBody
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.jsoup.select.Elements
import java.util.*
import java.util.regex.Pattern

/**
 * @author Jannik Bamberger (dev.jbamberger@gmail.com)
 */
internal object VPlanParser {

    private val REGEX_LAST_UPDATE = "Stand: ..\\...\\..... ..:..".toRegex()
    private val PATTERN_OMITTED = "entfall|x".toRegex()

    internal enum class Columns(pattern: String) {
        NR("vertretungsnummer|vtr-nr\\."),
        KIND("art|entfall"),
        HOUR("stunde"),
        CLASS("klasse(n|\\(n\\))?"),
        TEACHER("\\(?lehrer\\)?"),
        ROOM("\\(?raum\\)?"),
        SUBJECT("\\(?fach\\)?"),
        SUBST_FROM("vertr.\\s+von"),
        SUBST_TO("\\(le\\.\\) nach"),
        CONTENT("(vertretungs)?-?text");

        val regex = pattern.toRegex(RegexOption.IGNORE_CASE)
    }

    @Throws(ParseException::class)
    internal fun parseVPlanDay(body: ResponseBody): VPlanDay {
        val docString = readWithEncoding(body)
        val doc = Jsoup.parse(docString)
        return VPlanDay(
            VPlanHeader(readDayAndDate(doc), readLastUpdated(docString), readMotdTable(doc)),
            readVPlanTable(doc)
        )
    }

    @VisibleForTesting
    internal fun readWithEncoding(body: ResponseBody): String {
        val data = body.bytes()

        body.contentType()?.charset()?.let {
            return String(data, it)
        }

        val dataDefaultEncoded = String(data)
        val matcher = Pattern
            .compile("<meta\\s+http-equiv=\"Content-Type\"\\s+content=\"([^\"]*)\"\\s*/?>")
            .matcher(dataDefaultEncoded)

        while (matcher.find()) {
            matcher.group(1)!!.toMediaTypeOrNull()?.charset().let {
                if (it != null) return String(data, it)
            }
        }
        return dataDefaultEncoded
    }

    @VisibleForTesting
    @Throws(ParseException::class)
    internal fun readLastUpdated(html: String) = REGEX_LAST_UPDATE.find(html)?.value
        ?: throw ParseException("Could not find lastUpdated value in the document.")

    @VisibleForTesting
    @Throws(ParseException::class)
    internal fun readDayAndDate(vplanDoc: Document) = vplanDoc
        .getElementsByClass("mon_title")
        .singleOrNull()
        ?.allElements
        ?.firstOrNull()
        ?.text()
        ?: throw ParseException("Could not parse VPlan date and day.")

    @VisibleForTesting
    @Throws(ParseException::class)
    internal fun readMotdTable(doc: Document) = doc
        .getElementsByClass("info")
        .select("tr")
        .map { it.select("td") }
        .filter(Elements::isNotEmpty)
        .joinToString(separator = "<br>") { cells ->
            cells.joinToString(separator = " | ") { cell ->
                cell.html()
                    .split("<br>")
                    .joinToString(separator = "<br>", transform = String::trim)
            }
//                    if (it.first().text().toLowerCase(Locale.ROOT).contains("unterrichtsfrei")) {
//                        "<font color=#FF5252>$content</font>"
//                    } else {
//                        content
//                    }
        }


    @VisibleForTesting
    @Throws(ParseException::class)
    internal fun createHeaderColumnMapping(header: Element): Map<Columns, Int> {
        if (header.tagName() != "tr") {
            throw ParseException("First VPlan row is not the header. tag=${header.tagName()}")
        }

        return header.children().mapIndexed { index, col ->
            if (col.tagName() != "th") {
                throw ParseException("Header contained column with wrong tagName=${col.tagName()}.")
            }

            val colTitle = col.html().trim().lowercase(Locale.ROOT)
            val colName = Columns.values()
                .singleOrNull { it.regex.matches(colTitle) }
                ?: throw ParseException("Header column could not be assigned uniquely. Title=$colTitle")
            return@mapIndexed colName to index
        }.toMap()
    }

    @VisibleForTesting
    @Throws(ParseException::class)
    internal fun readVPlanTable(doc: Document): List<VPlanRow> {
        val rows = doc.getElementsByTag("table")
            .singleOrNull { it.hasClass("mon_list") }
            ?.getElementsByTag("tbody")
            ?.singleOrNull()
            ?.children()
            ?: throw ParseException("Couldn't locate VPlan table.")

        if (rows.isEmpty()) {
            throw ParseException("The VPlan table has no header")
        }

        if (rows.size == 1) {
            val firstRow = rows.removeAt(0)
            firstRow.text().contains("keine\\s+vertretungen".toRegex(RegexOption.IGNORE_CASE))
            return emptyList()
        }

        val colMap = createHeaderColumnMapping(rows.removeAt(0))

        return rows.map { readVPlanTableCells(it.children(), colMap) }
    }

    @VisibleForTesting
    @Throws(ParseException::class)
    internal fun readVPlanTableCells(cells: Elements, colMap: Map<Columns, Int>): VPlanRow {
        fun read(element: Element) = (element.getElementsByTag("span")
            .singleOrNull()
            ?.html() ?: element.text())
            .replace("&nbsp;", " ")
            .trim()

        fun resolveIndex(colName: Columns) = colMap[colName]
            ?: throw ParseException("could not resolve column name $colName")

        fun resolve(colName: Columns) = read(cells[resolveIndex(colName)])

        fun resolveOptional(colName: Columns) = try {
            resolve(colName)
        } catch (e: ParseException) {
            null
        }

        if (cells.size >= 6) {

            val isOmitted = cells[resolveIndex(KIND)].text()
                .lowercase(Locale.ROOT)
                .matches(PATTERN_OMITTED)
            val isNew = cells[resolveIndex(CLASS)]
                .attr("style")
                .matches("background-color: #00[Ff][Ff]00".toRegex())

            return VPlanRow(
                resolve(SUBJECT),
                isOmitted,
                resolve(HOUR),
                resolve(ROOM),
                resolve(CONTENT),
                resolve(CLASS),
                resolve(KIND).let { if (it.lowercase(Locale.ROOT) == "x") "entfall" else it },
                isNew,
                resolveOptional(NR),
                resolveOptional(TEACHER),
                resolveOptional(SUBST_FROM),
                resolveOptional(SUBST_TO)
            )
        }
        throw ParseException("Could not parse row, invalid format.")
    }
}

